pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -2013,3 +2015,43 @@ func LoadReleaseImageVerifier(config *rest.Config) (verify.Interface, error) {
 	}
 	return verify.NewFromManifests([]manifest.Manifest{m}, sigstore.NewCachedHTTPClientConstructor(sigstore.DefaultClient, nil).HTTPClient)
 }
+
+// IsErrorUpdateEvent returns true when the update event is from
+// error state in clusterdeployment.
+func IsErrorUpdateEvent(evt event.UpdateEvent) bool {
+	new, ok := evt.ObjectNew.(*hivev1.ClusterDeployment)
+	if !ok {
+		return false
+	}
+	if len(new.Status.Conditions) == 0 {
+		return false
+	}
+
+	old, ok := evt.ObjectOld.(*hivev1.ClusterDeployment)
+	if !ok {
+		return false
+	}
+
+	for _, cond := range new.Status.Conditions {
+		if controllerutils.IsConditionInDesiredState(cond) {
+			continue
+		}
+
+		oldcond := controllerutils.FindClusterDeploymentCondition(old.Status.Conditions, cond.Type)
+		if oldcond == nil {
+			return true // newly added condition in failed state
+		}
+
+		if controllerutils.IsConditionInDesiredState(*oldcond) {
+			return true // newly Failed condition
+		}
+
+		if cond.Message != oldcond.Message ||
+			cond.Reason != oldcond.Reason {
+			return true // already failing but change in error reported
+		}
```

Comment:
We have something like this that we pass on as a criteria when we try to [update conditions](https://github.com/openshift/hive/blob/4376099acc9c78577b1609347655c98bc98def28/pkg/controller/utils/conditions.go#L28-L33)
This takes away the logic from the setting of conditions
Is there a way that we have the `UpdateConditionAlways`, `UpdateConditionNever`, `UpdateConditionIfReasonOrMessageChange` logic in only 1 place?

Reply:
> Is there a way that we have the UpdateConditionAlways, UpdateConditionNever, UpdateConditionIfReasonOrMessageChange logic in only 1 place?

that goal for update criteria is different from what we want to do here. updating conditions requires allowing caller to decide when to update whereas here we are calculating if the change in conditions was due to error. Using a function like UpdateConditionIfReasonOrMessageChange here creates more confusion than clarity.